parquet_derive/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+        },
+        syn::GenericArgument::Const(_) => unimplemented!("generic argument const"),
+      }
+    },
+    PathArguments::Parenthesized(_) => unimplemented!("parenthesized"),
+  };
+
+  if &seg.ident.to_string()[..] == "Option" {
+    if let Some((ident2, _lifetime2, _is_option2)) = second_level {
+      (ident2, true, None)
+    } else {
+      unimplemented!("I couldn't parse what was inside of option")
+    }
+  } else {
+    (seg.ident.clone(), false, Some(seg.arguments.clone()))
+  }
+}
+
+impl FieldInfo {
+  fn from(f: &Field) -> Self {
+    let (field_type, field_lifetime, is_option) = match &f.ty {
+      Type::Slice(_) => unimplemented!("unsupported type: Slice"),
+      Type::Array(_) => unimplemented!("unsupported type: Array"),
+      Type::Ptr(_) => unimplemented!("unsupported type: Ptr"),
+      Type::Reference(ref tr) => {
+        //                unimplemented!("lifetime: {:#?}", lifetime);
+        extract_type_reference_info(tr)
+      },
+      Type::BareFn(_) => unimplemented!("unsupported type: BareFn"),
+      Type::Never(_) => unimplemented!("unsupported type: Never"),
+      Type::Tuple(_) => unimplemented!("unsupported type: Tuple"),
+      Type::Path(tp) => {
+        let (ident, is_option, _) = extract_path_info(&tp);
+        (ident, None, is_option)
+      },
+      Type::TraitObject(_) => unimplemented!("unsupported type: TraitObject"),
+      Type::ImplTrait(_) => unimplemented!("unsupported type: ImplTrait"),
+      Type::Paren(_) => unimplemented!("unsupported type: Paren"),
+      Type::Group(_) => unimplemented!("unsupported type: Group"),
+      Type::Infer(_) => unimplemented!("unsupported type: Infer"),
+      Type::Macro(_) => unimplemented!("unsupported type: Macro"),
+      Type::Verbatim(_) => unimplemented!("unsupported type: Verbatim"),
+    };
+
+    let column_writer_variant = match &field_type.to_string()[..] {
+      "bool" => quote! { parquet::column::writer::ColumnWriter::BoolColumnWriter },
+      "str" | "String" => {
+        quote! { parquet::column::writer::ColumnWriter::ByteArrayColumnWriter }
+      },
+      "i32" => quote! { parquet::column::writer::ColumnWriter::Int32ColumnWriter },
+      "i64" => quote! { parquet::column::writer::ColumnWriter::Int64ColumnWriter },
+      "f32" => quote! { parquet::column::writer::ColumnWriter::FloatColumnWriter },
+      "f64" => quote! { parquet::column::writer::ColumnWriter::DoubleColumnWriter },
+      o => unimplemented!("don't know {} for {:#?}", o, f),
+    };
+
+    FieldInfo {
+      field_name: f.ident.clone().expect("must be a named field"),
+      field_lifetime,
+      field_type,
+      is_option,
+      column_writer_variant,
+    }
+  }
+
+  pub fn to_writer_snippet(&self) -> proc_macro2::TokenStream {
+    let field_name = self.field_name.clone();
+    let field_type = self.field_type.clone();
+    let column_writer_variant = self.column_writer_variant.clone();
+    let is_option = self.is_option;
```

Comment:
My main concern is that Option<i32> of values [1, 2, 3, 4, 5, 6] will be encoded differently compared to i32 [1, 2, 3, 4, 5, 6], because of the definition levels. I think we should check if there are nulls first and then decide on what option to choose.

Reply:
Great point.